e2e/nomostest/portforwarder/port_forwarder.go

```diff
@@ -233,24 +223,80 @@ func (pf *PortForwarder) start() error {
 	// Restart the port forward if the process exits prematurely (e.g. due to pod eviction)
 	go func() {
 		for {
-			if err := pf.watcher.WatchForNotFound(kinds.Pod(), pf.pod, pf.ns, testwatcher.WatchTimeout(0), testwatcher.WatchContext(pf.ctx)); err != nil {
-				if errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
-					// stop retrying, the subprocess will be killed with the context.
-					return
-				}
-				// Watch automatically retries, so this shouldn't happen unless
-				// there's a terminal error.
-				pf.logger.Infof("Watch exited, restarting port-forward for %s/%s: %v", pf.ns, pf.deployment, err)
-			}
-			if err := pf.portForwardToDeployment(true); err != nil {
-				pf.logger.Info(err)
-				// this is a potentially fatal error since we failed to restart the port
-				// forward. we log the error and return in case the test is already done
-				// with the port forward. the test will otherwise error the next time it
-				// tries to use the local port.
+			if err := pf.waitAndRestart(); err != nil {
+				pf.logger.Infof("PortForwarder finished: %v", err)
 				return
 			}
 		}
 	}()
 	return nil
 }
+
+// waitAndRestart will block until seeing an indicator that the port-forward is
+// no longer healthy. If that occurs, it will restart the port-forward
+func (pf *PortForwarder) waitAndRestart() error {
+	// create an inner context for the watch so we can cancel threads between iterations
+	// the subprocess will be killed inside portForwardToDeployment
+	watchCtx, watchCancel := context.WithCancel(pf.ctx)
+	defer watchCancel()
+	// restart the port-forward if:
+	// - the subprocess is killed
+	// - the pod is restarted
+	restartChan := pf.signalOnFailure(
+		func() error {
+			return pf.cmd.Wait()
+		},
+		func() error {
+			return pf.watcher.WatchForNotFound(kinds.Pod(), pf.pod, pf.ns,
+				testwatcher.WatchTimeout(0), testwatcher.WatchContext(watchCtx))
+		},
+	)
+	select {
+	case <-pf.ctx.Done():
```

Comment:
can this also just be added to the signalOnFailure list? 
```
func() error {
   <-watchCtx.Done()
   return watchCtx.Err()
}
```

Then we only need to wait for <-restartChan (maybe rename doneCh) here, right? And we can check if the error is DeadlineExceeded or Cancelled to decide if we want to restart or not.

Reply:
done